ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
-				(const char *) needle_u8->val, needle_u8->len,
-				(const char *) offset_pointer);
+		while (search_pos < search_end) {
+			search_pos += (byte_len = mblen_table[*search_pos]);
 		}
-	}
 
-	if (found_pos) {
-		result = mbfl_pointer_to_offset_utf8(haystack_u8->val, (const unsigned char *) found_pos);
-	}
+		if (search_pos > search_end) {
+			search_pos -= byte_len;
+		}
 
-out:
-	if (haystack_u8 == &_haystack_u8) {
-		mbfl_string_clear(&_haystack_u8);
+		*length = search_pos - *start;
 	}
-	if (needle_u8 == &_needle_u8) {
-		mbfl_string_clear(&_needle_u8);
-	}
-	return result;
 }
 
-/*
- *  substr_count
- */
-
-size_t
-mbfl_substr_count(
-    mbfl_string *haystack,
-    mbfl_string *needle
-   )
-{
-	size_t n, result = 0;
-	unsigned char *p;
-	mbfl_convert_filter *filter;
-	struct collector_strpos_data pc;
-
-	/* needle is converted into wchar */
-	mbfl_wchar_device_init(&pc.needle);
-	filter = mbfl_convert_filter_new(
-	  needle->encoding,
-	  &mbfl_encoding_wchar,
-	  mbfl_wchar_device_output, 0, &pc.needle);
-	if (filter == NULL) {
-		return MBFL_ERROR_ENCODING;
-	}
-	mbfl_convert_filter_feed_string(filter, needle->val, needle->len);
-	mbfl_convert_filter_flush(filter);
-	mbfl_convert_filter_delete(filter);
-	pc.needle_len = pc.needle.pos;
-	if (pc.needle.buffer == NULL) {
-		return MBFL_ERROR_ENCODING;
-	}
-	if (pc.needle_len == 0) {
-		mbfl_wchar_device_clear(&pc.needle);
-		return MBFL_ERROR_EMPTY;
-	}
-	/* initialize filter and collector data */
-	filter = mbfl_convert_filter_new(
-	  haystack->encoding,
-	  &mbfl_encoding_wchar,
-	  collector_strpos, 0, &pc);
-	if (filter == NULL) {
-		mbfl_wchar_device_clear(&pc.needle);
-		return MBFL_ERROR_ENCODING;
-	}
-	pc.start = 0;
-	pc.output = 0;
-	pc.needle_pos = 0;
-	pc.found_pos = 0;
-	pc.matched_pos = MBFL_ERROR_NOT_FOUND;
-
-	/* feed data */
-	p = haystack->val;
-	n = haystack->len;
-	if (p != NULL) {
-		while (n > 0) {
-			if ((*filter->filter_function)(*p++, filter) < 0) {
-				pc.matched_pos = MBFL_ERROR_ENCODING;
-				break;
-			}
-			if (pc.matched_pos != MBFL_ERROR_NOT_FOUND) {
-				++result;
-				pc.matched_pos = MBFL_ERROR_NOT_FOUND;
-				pc.needle_pos = 0;
-			}
-			n--;
-		}
-	}
-	mbfl_convert_filter_flush(filter);
-	mbfl_convert_filter_delete(filter);
-	mbfl_wchar_device_clear(&pc.needle);
-
-	return result;
-}
-
-/*
- *  substr
- */
-struct collector_substr_data {
-	mbfl_convert_filter *next_filter;
-	size_t start;
-	size_t stop;
-	size_t output;
-};
-
-static int
-collector_substr(int c, void* data)
-{
-	struct collector_substr_data *pc = (struct collector_substr_data*)data;
-
-	if (pc->output >= pc->stop) {
-		return -1;
-	}
-
-	if (pc->output >= pc->start) {
-		(*pc->next_filter->filter_function)(c, pc->next_filter);
-	}
-
-	pc->output++;
-
-	return c;
-}
-
-mbfl_string *
-mbfl_substr(
-    mbfl_string *string,
-    mbfl_string *result,
-    size_t from,
-    size_t length)
+mbfl_string *mbfl_strcut(mbfl_string *string, mbfl_string *result, size_t from, size_t length)
```

Comment:
Because I was confused about this, maybe worth noting that `mbfl_strcut` only accepts a positive length. Caller is responsible for converting negative lengths.

Reply:
Added a comment.